src/dfx/src/actors/pocketic_proxy.rs

```diff
@@ -0,0 +1,361 @@
+        }
+    }
+
+    fn stopping(&mut self, _ctx: &mut Self::Context) -> Running {
+        self.stop_pocketic_proxy();
+
+        Running::Stop
+    }
+}
+
+impl Handler<PortReadySignal> for PocketIcProxy {
+    type Result = ();
+
+    fn handle(&mut self, msg: PortReadySignal, _ctx: &mut Self::Context) {
+        debug!(
+            self.logger,
+            "replica ready on {}, so re/starting HTTP gateway", msg.url
+        );
+
+        self.stop_pocketic_proxy();
+
+        let replica_url = Url::parse(&msg.url).unwrap();
+
+        self.start_pocketic_proxy(replica_url)
+            .expect("Could not start PocketIC HTTP gateway");
+    }
+}
+
+impl Handler<Shutdown> for PocketIcProxy {
+    type Result = ResponseActFuture<Self, Result<(), ()>>;
+
+    fn handle(&mut self, _msg: Shutdown, _ctx: &mut Self::Context) -> Self::Result {
+        // This is just the example for ResponseActFuture but stopping the context
+        Box::pin(
+            async {}
+                .into_actor(self) // converts future to ActorFuture
+                .map(|_, _act, ctx| {
+                    ctx.stop();
+                    Ok(())
+                }),
+        )
+    }
+}
+
+fn pocketic_proxy_start_thread(
+    logger: Logger,
+    address: SocketAddr,
+    replica_url: Url,
+    pocketic_proxy_path: PathBuf,
+    pocketic_proxy_pid_path: PathBuf,
+    pocketic_proxy_port_path: PathBuf,
+    receiver: Receiver<()>,
+    verbose: bool,
+    domains: Vec<String>,
+) -> DfxResult<std::thread::JoinHandle<()>> {
+    let thread_handler = move || {
+        loop {
+            // Start the process, then wait for the file.
+
+            // form the pocket-ic command here similar to replica command
+            let mut cmd = std::process::Command::new(&pocketic_proxy_path);
+            if !verbose {
+                cmd.env("RUST_LOG", "error");
+            }
+
+            cmd.args(["--ttl", "2592000"]);
+            cmd.args(["--port-file".as_ref(), pocketic_proxy_port_path.as_os_str()]);
+            let tmp = tempdir().expect("Could not create temporary directory.");
+            let ready_file = tmp.path().join("ready");
+            cmd.args(["--ready-file".as_ref(), ready_file.as_os_str()]);
+            cmd.stdout(std::process::Stdio::inherit());
+            cmd.stderr(std::process::Stdio::inherit());
+            let last_start = std::time::Instant::now();
+            debug!(logger, "Starting pocket-ic gateway...");
+            let mut child = cmd.spawn().expect("Could not start pocket-ic gateway.");
+
+            std::fs::write(&pocketic_proxy_pid_path, "")
+                .expect("Could not write to pocketic-proxy-pid file.");
+            std::fs::write(&pocketic_proxy_pid_path, child.id().to_string())
+                .expect("Could not write to pocketic-proxy-pid file.");
+            let port =
+                PocketIcProxy::wait_for_ready(&pocketic_proxy_port_path, &ready_file).unwrap();
```

Comment:
Could we also check for the kill signals here so that we aren't hanging until the maximum number of retries in case the PocketIC server has been CTRL-C'ed?